contrib/babelfishpg_common/src/datetime.c

```diff
@@ -46,6 +47,361 @@ PG_FUNCTION_INFO_V1(datetime_mi_datetime);
 void		CheckDatetimeRange(const Timestamp time);
 void		CheckDatetimePrecision(fsec_t fsec);
 
+/*
+ * This function checks the regex in case of the text month.
+ * We will be checking with the combination of date regex's and
+ * time regex's.
+ */
+bool
+check_regex_for_text_month(char *str, DateTimeContext context)
+{
+	regex_t regex;
+	char curr_regex[200];
+    	int status, i, j;
+	const char *regex_time_offset = "(((\\+|\\-)[0-9]{1,2}:[0-9]{1,2})|Z)?";
+
+	for (i = 0; i < NUM_DATE_REGEXES; i++) {
+		/*
+		 * check for just date syntax
+		 */
+		strcpy(curr_regex, "^");
+        	strcat(curr_regex, date_regexes[i]);
+        	strcat(curr_regex, "$");
+    		status = regcomp(&regex, curr_regex, REG_EXTENDED);
+    		status = regexec(&regex, str, 0, NULL, 0);
+    		regfree(&regex);
+
+		if (status == 0)
+        		return 1;
+
+        	for (j = 0; j < NUM_TIME_REGEXES; j++) {
+			/*
+			 * check for just time syntax
+			 */
+			strcpy(curr_regex, "^");
+        		strcat(curr_regex, time_regexes[j]);
+			if (context == DATE_TIME_OFFSET)
+			{
+				strcat(curr_regex, "\\s*");
+            			strcat(curr_regex, regex_time_offset);
+			}
+        		strcat(curr_regex, "$");
+            		status = regcomp(&regex, curr_regex, REG_EXTENDED);
+            		status = regexec(&regex, str, 0, NULL, 0);
+            		regfree(&regex);
+
+			if (status == 0)
+                		return 1;
+
+	    		/*
+			 * check for the combination of date and time
+			 */
+			strcpy(curr_regex, "^");
+			strcat(curr_regex, date_regexes[i]);
+            		strcat(curr_regex, "\\s+");
+            		strcat(curr_regex, time_regexes[j]);
+			if (context == DATE_TIME_OFFSET)
+			{
+				strcat(curr_regex, "\\s*");
+            			strcat(curr_regex, regex_time_offset);
+			}
+			strcat(curr_regex, "$");
+            		status = regcomp(&regex, curr_regex, REG_EXTENDED);
+            		status = regexec(&regex, str, 0, NULL, 0);
+            		regfree(&regex);
+
+            		if (status == 0)
+                		return 1;
+        	}
+    	}
+
+	return 0;
+}
+
+/*
+ * This function cleans up the input string by removing
+ * unnecessary spaces between the fields.
+ */
+char*
+clean_input_str(char *str, bool *contains_extra_spaces, DateTimeContext context)
+{
+	char *result = (char *) palloc(MAXDATELEN);
+	int i = 0, j = 0;
+	int last_non_space = -1;
+	int num_colons = 0;
+
+	while (str[i] != '\0')
+	{
+		/*
+		 * Remove leading spaces
+		 */
+		while (str[i] != '\0' && isspace(str[i]))
+		{
+			if (i != 0)
+				*contains_extra_spaces = true;
+			i++;
+		}
+
+		if (str[i] == '\0')
+			break;
+
+		if (context == DATE_TIME_OFFSET && str[i] == ':')
+			num_colons++;
+
+		/*
+		 * Modify DATE delimiters to '.' as in DATETIME multiple delimiters can
+		 * be passed for DATE field.
+		 */
+		if (context == DATE_TIME && (str[i] == '/' || str[i] == '-' || str[i] == '.'))
+		{
+			result[j] = '.';
+			j++;
```

Comment:
We must check if the value of `j` is within `MAXDATELEN` or otherwise we might write beyond the allocated length. Also, let's add test cases to test this which includes too large dates with length greater than MAXDATELEN.